modules/core/src/main/java/org/apache/ignite/internal/managers/tracing/GridTracingManager.java

```diff
@@ -179,8 +211,127 @@ private Span enrichWithLocalNodeParameters(@Nullable Span span) {
     }
 
     /** {@inheritDoc} */
-    @Override public Span create(@NotNull SpanType spanType, @Nullable byte[] serializedSpan) {
-        return enrichWithLocalNodeParameters(getSpi().create(spanType, serializedSpan));
+    @Override public Span create(@NotNull SpanType spanType, @Nullable byte[] serializedParentSpan) {
+        // 1 byte: special flags;
+        // 1 bytes: spi type;
+        // 2 bytes: major protocol version;
+        // 2 bytes: minor protocol version;
+        // 4 bytes: spi specific serialized span length;
+        // n bytes: spi specific serializes span length;
+        // 4 bytes: span type
+        // 4 bytes included scopes size;
+        // 2 * included scopes size: included scopes items one by one;
+
+        Span span;
+
+        try {
+            if (serializedParentSpan == null || serializedParentSpan == NoopTracing.NOOP_SERIALIZED_SPAN)
+                return create(spanType, NoopSpan.INSTANCE);
+
+            // First byte of the serializedSpan is reserved for special flags - it's not used right now.
+
+            // Deserialize and compare spi types. If they don't match (span was serialized with another spi) then
+            // propagate serializedSpan as DeferredSpan.
+            if (serializedParentSpan[1] != getSpi().type().index())
+                return new DeferredSpan(serializedParentSpan);
+
+            // Deserialize and check major protocol version,
+            // cause protocol might be incompatible in case of different protocol versions -
+            // propagate serializedSpan as DeferredSpan.
+            if (serializedParentSpan[2] != MAJOR_PROTOCOL_VERSION)
+                return new DeferredSpan(serializedParentSpan);
+
+            // Deserialize and check minor protocol version.
+            // within the scope of the same major protocol version, protocol should be backwards compatible
+            byte minProtoVer = serializedParentSpan[3];
+
+            // Deserialize spi specific span size.
+            int spiSpecificSpanSize = bytesToInt(
+                Arrays.copyOfRange(
+                    serializedParentSpan,
+                    4,
+                    8),
+                0);
+
+            SpanType parentSpanType = null;
+
+            Set<Scope> includedScopes = new HashSet<>();
+
+            // Fall through.
+            switch (minProtoVer) {
+                case 0 : {
+                    // Deserialize parent span type.
+                    parentSpanType = SpanType.fromIndex(
+                        bytesToInt(
+                            Arrays.copyOfRange(
+                                serializedParentSpan,
+                                8 + spiSpecificSpanSize,
+                                12 + spiSpecificSpanSize),
```

Comment:
It would be a bit safer to create constants for previous items lengths and calculate prefix size with "+"
See: PageIO

Reply:
Applicable to all next code lines with hardcoded constants